File: UEMPC.cpp
#include "UEMPC.h"

int main(int argc, char **argv) {
    int counter = 0;

    zmq::context_t ctx;
    zmq::socket_t sock(ctx, zmq::socket_type::sub);
    sock.connect("tcp://127.0.0.1:6666");
    sock.set(zmq::sockopt::subscribe, "");

    std::this_thread::sleep_for(std::chrono::milliseconds(1000));

    while (true) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));

        zmq::message_t message;
        (void) sock.recv(message);

        MSG msg;
        msgpack::unpacked result;
        std::stringstream sbuf;
        sbuf << message.to_string();
        std::size_t off = 0;
        msgpack::unpack(result, sbuf.str().data(), sbuf.str().size(), off);
        result.get().convert(msg);

        try {
            if (msg.id == MSG_ID::STATE_PLAYER) {
                STATE_PLAYER state_player;
                msg.msg.convert(state_player);
                std::cout << " [client] " << counter << " " << unsigned(state_player.id) << " " << state_player.name << " " << unsigned(state_player.level) << "\n";
            } else if (msg.id == MSG_ID::STATE_NPC) {
                STATE_NPC state_npc;
                msg.msg.convert(state_npc);
                std::cout << " [client] " << counter << " " << unsigned(state_npc.id) << " " << state_npc.name << " " << unsigned(state_npc.health) << " " << unsigned(state_npc.strength) << "\n";
            }
        } catch (const std::exception &e) {
            std::cout << " [client] " << counter << " " << e.what() << "\n";
        }

        counter++;
    }
}

File: UEMPC.py
import sys
import time
import threading
import random
import zmq
import msgpack
import UEMP


def subscriber():
    """This function is used to create a subscriber that receives messages from the publisher"""

    # create ZeroMQ client
    context = zmq.Context()
    socket = context.socket(zmq.SUB)
    socket.connect("tcp://127.0.0.1:6666")
    socket.setsockopt(zmq.SUBSCRIBE, b"")

    # infinite loop
    while True:
        # receive the message
        serialized = socket.recv()

        # deserialize the message
        packed = msgpack.unpackb(serialized)
        print(packed)
        message = UEMP.MSG()
        message.unpack(packed)

        # print the message
        print("Client:", round(time.time(), 3), "recv", message)


def pusher(instance):
    """This function is used to create a pusher that sends messages to the puller"""

    # create ZeroMQ client
    context = zmq.Context()
    socket = context.socket(zmq.PUSH)
    socket.connect("tcp://127.0.0.1:6667")

    # infinite loop
    while True:
        # create a message object
        message = UEMP.MSG(UEMP.MSG_ID.STATE_PLAYER,
                           UEMP.STATE_PLAYER(instance, f"Player {instance}", random.randint(0, 100)))

        # print the message
        # print("Client:", round(time.time(), 3), "send", message)

        # serialize the object with its method
        packed = message.pack()
        serialized = msgpack.packb(packed)

        # send the message with no wait
        socket.send(serialized, zmq.DONTWAIT)

        # wait for some time
        time.sleep(0.1)


def main():
    """This function is used to create the subscriber and the pusher"""

    # get input arguments
    instance = sys.argv[1]

    # create the subscriber
    subscriber_thread = threading.Thread(target=subscriber)
    subscriber_thread.start()

    # create the pusher
    pusher_thread = threading.Thread(target=pusher, args=(instance,))
    pusher_thread.start()

    # wait for the threads to finish
    subscriber_thread.join()
    pusher_thread.join()


if __name__ == "__main__":
    main()
